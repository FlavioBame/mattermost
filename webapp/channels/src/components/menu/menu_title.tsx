// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {
    ReactNode,
} from 'react';
import React from 'react';
import styled from 'styled-components';

export interface Props {
    children: ReactNode;
}

export function MenuTitle(props: Props) {
    const {
        children,
    } = props;

    return (
        <Title>
            {children}
        </Title>
    );
}

const Title = styled.h4`
    font-size: 12px;
    font-weight: 600;
    line-height: 16px;
    color: rgba(var(--center-channel-color-rgb), 0.56);
    text-transform: uppercase;
    padding: 6px 20px;
    margin: 0;
`;
