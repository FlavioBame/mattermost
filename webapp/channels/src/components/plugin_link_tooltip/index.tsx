// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {
    useFloating,
    autoUpdate,
    safePolygon,
    useHover,
    useDismiss,
    useInteractions,
    FloatingPortal,
    autoPlacement,
    inline,
    useTransitionStyles,
    FloatingOverlay,
    FloatingFocusManager,
    arrow,
    offset,
    FloatingArrow,
} from '@floating-ui/react';
import React, {useRef, useState} from 'react';
import type {ReactElement} from 'react';

import Pluggable from 'plugins/pluggable';
import {RootHtmlPortalId, OverlaysTimings, OverlayArrow} from 'utils/constants';

import './plugin_link_tooltip.scss';

interface Props {
    href: string;
    attributeDataHashtag?: string;
    attributeDataLink?: string;
    attributeDataChannelMention?: string;
    children: ReactElement;
}

function PluginLinkTooltip(props: Props) {
    const [isOpen, setOpen] = useState(false);

    const arrowRef = useRef(null);

    const {refs: {setReference, setFloating}, floatingStyles, context: floatingContext} = useFloating({
        open: isOpen,
        onOpenChange: setOpen,
        whileElementsMounted: autoUpdate,
        middleware: [
            offset(OverlayArrow.OFFSET),
            inline(),
            autoPlacement({
                allowedPlacements: ['top', 'bottom'],
            }),
            arrow({
                element: arrowRef,
            }),
        ],
    });

    const {isMounted, styles: transitionStyles} = useTransitionStyles(floatingContext, {
        duration: {
            open: OverlaysTimings.FADE_IN_DURATION,
            close: OverlaysTimings.FADE_OUT_DURATION,
        },
        initial: {
            opacity: 0,
        },
        common: {
            opacity: 1,
        },
    });

    const combinedFloatingStyles = Object.assign({}, floatingStyles, transitionStyles);

    const hoverInteractions = useHover(floatingContext, {
        restMs: OverlaysTimings.CURSOR_REST_TIME_BEFORE_OPEN,
        move: false,
        handleClose: safePolygon({
            requireIntent: false,
            blockPointerEvents: true,
        }),
    });
    const dismissInteraction = useDismiss(floatingContext);

    const {getReferenceProps, getFloatingProps} = useInteractions([
        hoverInteractions,
        dismissInteraction,
    ]);

    return (
        <>
            <span
                ref={setReference}
                data-hashtag={props.attributeDataHashtag}
                data-link={props.attributeDataLink}
                data-channel-mention={props.attributeDataChannelMention}
                {...getReferenceProps()}
            >
                {props.children}
            </span>
            {isMounted && (
                <FloatingPortal id={RootHtmlPortalId}>
                    <FloatingOverlay className='plugin-link-tooltip-floating-overlay'>
                        <FloatingFocusManager context={floatingContext}>
                            <div
                                ref={setFloating}
                                style={combinedFloatingStyles}
                                className='plugin-link-tooltip-container'
                                {...getFloatingProps()}
                            >
                                <Pluggable
                                    href={props.href}
                                    show={true}
                                    pluggableName='LinkTooltip'
                                />
                                <FloatingArrow
                                    ref={arrowRef}
                                    context={floatingContext}
                                    width={OverlayArrow.WIDTH}
                                    height={OverlayArrow.HEIGHT}
                                    className='plugin-link-tooltip-arrow'

                                    // Shift so the border of base of arrow triangle merges with the popover container
                                    style={{transform: 'translateY(-1px'}}
                                />
                            </div>
                        </FloatingFocusManager>
                    </FloatingOverlay>
                </FloatingPortal>
            )}
        </>
    );
}

export default PluginLinkTooltip;
