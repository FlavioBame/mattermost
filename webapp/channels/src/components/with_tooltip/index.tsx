// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {Placement} from '@floating-ui/react';
import {
    useFloating,
    autoUpdate,
    offset,
    useHover,
    useFocus,
    useDismiss,
    useRole,
    useInteractions,
    arrow,
    FloatingPortal,
    useTransitionStyles,
    FloatingArrow,
    flip,
    useMergeRefs,
} from '@floating-ui/react';
import classNames from 'classnames';
import React, {useRef, useState, useMemo, cloneElement, isValidElement} from 'react';
import type {ReactElement, ReactNode} from 'react';
import type {MessageDescriptor} from 'react-intl';
import {defineMessage} from 'react-intl';

import {OverlayArrow, OverlaysTimings} from 'utils/constants';

import TooltipContent from './tooltip_content';
import type {ShortcutDefinition} from './tooltip_shortcut';

import './with_tooltip.scss';

export const ShortcutKeys = {
    alt: defineMessage({
        id: 'shortcuts.generic.alt',
        defaultMessage: 'Alt',
    }),
    cmd: '⌘',
    ctrl: defineMessage({
        id: 'shortcuts.generic.ctrl',
        defaultMessage: 'Ctrl',
    }),
    option: '⌥',
    shift: defineMessage({
        id: 'shortcuts.generic.shift',
        defaultMessage: 'Shift',
    }),
};

interface Props {
    title: string | ReactNode | MessageDescriptor;
    emoji?: string;
    isEmojiLarge?: boolean;
    hint?: string | ReactNode | MessageDescriptor;
    shortcut?: ShortcutDefinition;

    /**
     * Whether the tooltip should be vertical or horizontal, by default it is vertical
     * This doesn't always guarantee the tooltip will be vertical, it just determines the initial placement and fallback placements
    */
    isVertical?: boolean;
    tooltipContentContainerClassName?: string;
    disabled?: boolean;

    /**
    * @deprecated Do not use this except for special cases
    * Callback when the tooltip appears
   */
    onOpen?: () => void;
    children: ReactElement;
}

function WithTooltip({
    children,
    title,
    emoji,
    isEmojiLarge = false,
    hint,
    shortcut,
    isVertical = true,
    tooltipContentContainerClassName,
    onOpen,
    disabled,
}: Props) {
    const [open, setOpen] = useState(false);

    const arrowRef = useRef(null);

    function handleChange(open: boolean) {
        setOpen(open);

        if (onOpen && open) {
            onOpen();
        }
    }

    const placements = useMemo<{initial: Placement; fallback: Placement[]}>(() => {
        let initial: Placement;
        let fallback: Placement[];
        if (isVertical) {
            initial = 'top';
            fallback = ['bottom', 'right', 'left'];
        } else {
            initial = 'right';
            fallback = ['left', 'top', 'bottom'];
        }
        return {initial, fallback};
    }, [isVertical]);

    const {refs: {setReference, setFloating}, floatingStyles, context} = useFloating({
        open: disabled ? false : open,
        onOpenChange: handleChange,
        whileElementsMounted: autoUpdate,
        placement: placements.initial,
        middleware: [
            offset(OverlayArrow.OFFSET),
            flip({
                fallbackPlacements: placements.fallback,
            }),
            arrow({
                element: arrowRef,
            }),
        ],
    });

    const hover = useHover(context, {
        restMs: OverlaysTimings.CURSOR_REST_TIME_BEFORE_OPEN,
    });
    const focus = useFocus(context);
    const dismiss = useDismiss(context);
    const role = useRole(context, {role: 'tooltip'});

    const {getReferenceProps, getFloatingProps} = useInteractions([hover, focus, dismiss, role]);
    const {isMounted, styles: transitionStyles} = useTransitionStyles(context, {
        duration: {
            open: OverlaysTimings.FADE_IN_DURATION,
            close: OverlaysTimings.FADE_OUT_DURATION,
        },
        initial: {
            opacity: 0,
        },
        common: {
            opacity: 1,
        },
    });

    if (!isValidElement(children)) {
        // eslint-disable-next-line no-console
        console.error('Children must be a valid React element for WithTooltip');
    }

    const mergedRefs = useMergeRefs([(children as any)?.ref, setReference]);

    const trigger = cloneElement(children, {
        ...getReferenceProps({
            ref: mergedRefs,
            ...children.props,
        }),
    });

    return (
        <>
            {trigger}
            {isMounted && (
                <FloatingPortal
                    id='root-portal' // This is the global portal container id
                >
                    <div
                        className='tooltipContainer'
                        ref={setFloating}
                        style={floatingStyles}
                        {...getFloatingProps()}
                    >
                        <div
                            className={classNames('tooltipContentContainer', tooltipContentContainerClassName)}
                            style={transitionStyles}
                        >
                            <TooltipContent
                                title={title}
                                emoji={emoji}
                                isEmojiLarge={isEmojiLarge}
                                hint={hint}
                                shortcut={shortcut}
                            />
                            <FloatingArrow
                                ref={arrowRef}
                                context={context}
                                width={OverlayArrow.WIDTH}
                                height={OverlayArrow.HEIGHT}
                            />
                        </div>
                    </div>
                </FloatingPortal>
            )}
        </>
    );
}

export default WithTooltip;
