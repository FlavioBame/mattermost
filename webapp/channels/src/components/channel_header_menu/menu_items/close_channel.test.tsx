// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {shallow} from 'enzyme';
import React from 'react';

import * as Menu from 'components/menu';

import CloseChannel from './close_channel';

describe('components/ChannelHeaderDropdown/MenuItem.CloseChannel', () => {
    const baseProps = {
        goToLastViewedChannel: jest.fn(),
    };

    it('should match snapshot', () => {
        const wrapper = shallow(<CloseChannel {...baseProps}/>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should runs goToLastViewedChannel function on click', () => {
        const mockFunction = jest.fn();
        const props = {
            goToLastViewedChannel: mockFunction,
        };
        const wrapper = shallow(<CloseChannel {...props}/>);

        expect(wrapper.find(Menu.Item)).toBeDefined();
        wrapper.find(Menu.Item).simulate('click');
        expect(mockFunction).toHaveBeenCalled();
    });
});
