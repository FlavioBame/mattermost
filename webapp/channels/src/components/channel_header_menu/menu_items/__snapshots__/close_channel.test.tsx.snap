// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ChannelHeaderDropdown/MenuItem.CloseChannel should match snapshot 1`] = `
<MenuItem
  labels={
    <Memo(MemoizedFormattedMessage)
      defaultMessage="Close Channel"
      id="center_panel.archived.closeChannel"
    />
  }
  onClick={[MockFunction]}
/>
`;
